### DISCUSSION TITLE
Possible to get battery total capacity and usable capacity in kWh?

### DISCUSSION THREAD

**Discussion by andrewmooreio** [nathanmarlor/foxess_modbus#359] [General]
Hi,

Just wondering if there are any registers for total battery capacity or usable capacity in kWh from the BMS? Similar to the BMS kWh Remaining sensor.

I have 4x HV2600 batteries, which should give me a nominal capacity of 10.24 kWh.

When SoC hits 100%, my BMS kWh Remaining is approximately 9.7 kWh. This varies somewhat, it can be as low as 9.6 kWh or as high as 9.76 kWh. It would be good to have a reliable value.

Thanks!

**Comment by andrewmooreio** [comment]
The motivation behind this is I'd like to work out the kWh remaining to 100%, then have some logic which calculates how long it'll take to fully charge at the current charge rate.

**Reply by FozzieUK** [reply]
There isn't an 'as installed' total capacity register available on the system, however the kWh remaining is the BMS giving you it's very best estimate of 'actual' kWh remaining based on the information from the packs and any other parameters (temperature, cycles etc..) it has available. That's why it changes from time to time, it is recalculating on the basis of current best known balance state and is a good source for any sensor you want to determine usable capacity or runtime from.

For the test you want to do you might be better, using the SoC and the known maximum of the batteries (I have that stored in a helper input_number.battery_kwh_capacity) and then have a template sensor that calculates the time to full as the SoC (or kWh changes).

I have a few template battery sensors that give me the usable battery % (taking minsoc into account), I calculate available energy from that, and that also gives me an approx runtime based on todays average load.

For my charge current I don't calculate time to full charge but do control how the batteries are charged as they approach full - I use an automation that triggers each time the SoC changes, I charge at a fixed level below 80% and above that calculate the maximum allowable charge rate to give me a custom charge profile as it approaches 100% to avoid excess stress and help with absorption.

![image](https://github.com/nathanmarlor/foxess_modbus/assets/113460294/bc3d4ea4-66e8-461f-8a81-d2af97e393e8)

**Reply by calum-mcfarlane** [reply]
I've noticed my H1 setup will keep battering _all the amps_ into my x6 HV2600s even when the SoC is very high, so I'd be very interested in seeing how you do this (I can guess at the logic but my programming experience is limited to GFA Basic on the Atari ST, so I'm somewhat out of date...) 

As an aside, I can't help but think something like this would be an AMAZING feature for an integration like foxess_modbus to have, since it really should meaningfully extend the life of the batteries, which as we know are very energy intensive to produce (and expensive to buy!)

**Reply by canton7** [reply]
Interestingly, my bms trickle-charges the batteries above 95%. Given that this apparently changes with things like battery / firmware version I rather suspect foxess know what they're doing, and I would be strongly disinclined to push people towards doing something else.

**Reply by calum-mcfarlane** [reply]
There's a rabbit hole here about which batteries and BMS and what firmware versions, but very fair point. @FozzieUK, still intrqigued to see how you do it because my BMS is not so kindly disposed...

**Reply by FozzieUK** [reply]
@calum-mcfarlane We did quite a bit of work on this last year (for the Energy Management integration) as the BMS master/slave implementation really wasn’t good back then (it is light years better now).

LifePo4 batteries have a very flat voltage profile across 85% of their usable range, with a rapid fall below 10% minsoc, and rapid increase when nearing full - advanced LFP battery chargers will watch the individual voltage and cell temperatures, reducing charge current but holding voltage which avoids the extreme temperature rise you get across the SEI layers resistance when the voltage increases; this allows the maximum time for the charge to be absorbed safely.

The Foxess charger (in the inverter) will ramp the charge current down as it approaches full (the BMS controls it using the BMS charge rate register) - you normally see a stepped charge rate 4kw>1kw>500w>off. 
There is no voltage hold and that rises quickly at full which you will see if you look at the battery voltage and internal temperature. It does a reasonable job, in winter the warming of the batteries is a good thing but in summer it can add excessive heat stress and often the BMS cuts the charge early which is why you hear complaints their batteries stop charging at 90-95% randomly.

@nathanmarlor added a custom profile option in the Energy Management integration which ramped the charge current down during the last 10% of charge and it works really well for grid charging on Eco tariffs - the only downside was it didn’t control solar charge, and being a bit too assiduous I made a charge throttle automation that is agnostic of charge source and gives me a profiled charge with a very soft charge profile above 90%.

It’s mostly just a few helpers (last_battery_soc, max charge limit), an automation to trigger when maxsoc changes, and a jinja import (basically a big if/elif) which provides the charge multiplier % for the calculation, it then calls a script that writes the calculated rate to the inverters max charge current.
If I get a couple of spare hours over the next few days, I’ll write it up and pop it on the Foxesscommunity website.

**Reply by FozzieUK** [reply]
@calum-mcfarlane - ok it's over on the [community website](https://foxesscommunity.com/viewtopic.php?t=368) if you want to take a look 👍

**Comment by TonyM1958** [comment]
> Hi,
> 
> Just wondering if there are any registers for total battery capacity or usable capacity in kWh from the BMS? Similar to the BMS kWh Remaining sensor.
> 
> I have 4x HV2600 batteries, which should give me a nominal capacity of 10.24 kWh.
> 
> When SoC hits 100%, my BMS kWh Remaining is approximately 9.7 kWh. This varies somewhat, it can be as low as 9.6 kWh or as high as 9.76 kWh. It would be good to have a reliable value.
> 
> Thanks!

From what I have seen, this depends on the version of BMS you have - v1 tries to fill the batteries to 100% but v2 stops at 90%. So, if you have v2, you can expect the batteries to be full and SoC to read 100% when BMS_kWh_remaining = 10.24 * 0.9 = 9.2kWh. The fact you are getting 9.6 or 9.7kwh suggests to me that you have BMS v1 and there may be a cell balance issue that stops the BMS from getting to 10.24kWh?

**Reply by andrewmooreio** [reply]
> > Hi,
> 
> > 
> 
> > Just wondering if there are any registers for total battery capacity or usable capacity in kWh from the BMS? Similar to the BMS kWh Remaining sensor.
> 
> > 
> 
> > I have 4x HV2600 batteries, which should give me a nominal capacity of 10.24 kWh.
> 
> > 
> 
> > When SoC hits 100%, my BMS kWh Remaining is approximately 9.7 kWh. This varies somewhat, it can be as low as 9.6 kWh or as high as 9.76 kWh. It would be good to have a reliable value.
> 
> > 
> 
> > Thanks!
> 
> 
> 
> From what I have seen, this depends on the version of BMS you have - v1 tries to fill the batteries to 100% but v2 stops at 90%. So, if you have v2, you can expect the batteries to be full and SoC to read 100% when BMS_kWh_remaining = 10.24 * 0.9 = 9.2kWh. The fact you are getting 9.6 or 9.7kwh suggests to me that you have BMS v1 and there may be a cell balance issue that stops the BMS from getting to 10.24kWh?

Strange. I definitely have the V2 BMS. I've set up a Home Assistant sensor to calculate the approximate max SoC and it ranges from 9.64kWh to 9.85kWh. The environment the batteries are in is very warm, if that makes a difference.

**Reply by TonyM1958** [reply]
Always interesting to hear what other people see - more data points gives better understanding. Are all the batteries v2 also?

**Comment by FozzieUK** [comment]
> From what I have seen, this depends on the version of BMS you have - v1 tries to fill the batteries to 100% but v2 stops at 90%. So, if you have v2, you can expect the batteries to be full and SoC to read 100% when BMS_kWh_remaining = 10.24 * 0.9 = 9.2kWh. The fact you are getting 9.6 or 9.7kwh suggests to me that you have BMS v1 and there may be a cell balance issue that stops the BMS from getting to 10.24kWh?

@TonyM1958 the old V1 BMS used to charge to 99% but the latest firmware makes it the same as V2 BMS, both are now consistent and charge to 100%  - they both use same the same KwR remaining calculation.

@andrewmooreio It does sound like you have one or more packs slightly out of balance, the first check is do you have the latest V2 firmware installed? - BCU is 1.008, BMU is 1.07 (you can see these from the inverter panel select About, Batt Ver)
Each of the packs manages it's own cell balance, the BMS manages the pack balance - both of these are helped by occasionally running the pack to 10% and then recharging to 100%, preferably with the dc charge current set no higher than 5A which gives the balance circuits more time to do their job.

**Reply by TonyM1958** [reply]
@FozzieUK I think that's a slightly different thing, as I understand it. There was a problem where the v1 BMS SoC only got to 99% that was fixed recently so it went to 100% as part of the calibration process.

What I'm talking about here is the actual energy stored compared to the battery capacity when the BMS has calibrated and says SoC=100%. With my v2 BMS and v2 batteries, it only ever fills the battery to 90% of capacity and shows this as 100% SoC. With other v1 BMS and v1 batteries, as reported to me, it tries to fill to 100% of capacity and shows this as 100% SoC ( I did wonder if that was why v1 often stopped at 99% to avoid battery damage, and if the recent update tweaked the calibration so 99% capacity shows as 100% SoC).

@andrewmooreio if you are getting more than 90% of the battery capacity at 100% SoC with a v2 BMS and v2 batteries, you are actually getting more usable capacity than me. My cell imbalance usually settles at around 0.2% with a cell high of 3.362v and a cell lows of 3.355 so I don't think I have a balance issue... I did wonder if you might be getting more if you had a v2 BMS with a mix of v1 and v2 batteries?

There are various calculations I do around battery capacity and health in my integration. Here's a quick summary of the template sensors I calculate, with examples based on my 6 x HV2600 v2:

+ Battery Specification: the data sheet battery kWh (15.36kWh)
+ Battery Capacity: the value of bms_kwh_remaining when SoC=100% (13.82 kWh)
+ Battery SoH: the ratio of Battery Capacity / Battery Specification. For v2 BMS, this is typically 90% but for v1 it can go up to 100%
+ Battery life: the remaining battery life based on extrapolating the number of days since installation and the BMS cycle count to predict when the battery will get to 6,000 cycles. This was around 25 years in Spring but is now up to 27 years.
+ Battery Energy Total: this is sum of Charge Energy Total and Discharge Energy Total - Fox say the warranty is limited to 6MWh per battery for HV2600. I have 6 batteries, so my warranty is void when I get to 36MWh. I am currently at 2.3MWh after 6 months of use, so my warranty is likely to be void around 8 years after installation...
+ Battery Energy per cycle: Battery Energy Total / BMS Cycle Count - an indicator of what the BMS is looking at when counting cycles and how hard the batteries are being worked. This was around 19kwh / cycle in Spring but is up to 22kwh / cycle now.

![image](https://github.com/nathanmarlor/foxess_modbus/assets/63789168/eb599fd5-106b-4c9a-acc8-3ef319aca13a)

The difference between Spring and now is the reduction in force charging and max charge current over the summer. I expect some of the indicators will drop in Autumn / Winter - but at least I will have a feel or how hard / soft I am being on the batteries.

There is a view that over charging is bad for the long term battery life as it can lead to thermal issues and cell imbalance. I think this may be the reason Fox lowered the level they push the battery up to when charging with the v2 BMS?

**Reply by FozzieUK** [reply]
@TonyM1958 interesting, thanks for that clarification - I suppose it does make sense to remove the 10% that cannot be used from the calculation to make it a more realistic 'usable capacity'. 

The v1 batteries report their individual soc's, kwr calculation to the v1 BMS (see screenshot of my packs) but if the v2 BMS is using a different calculation base I suspect we may be getting close to the source of the capacity issues and soc drift in the mixed v1/v2 packs - there is an R&D release just out supposedly that nails this, fingers crossed  🙄 

![IMG_7917](https://github.com/nathanmarlor/foxess_modbus/assets/113460294/543bdb65-1dc1-4cd0-a25f-06a7206480fd)

**Reply by calum-mcfarlane** [reply]
If I understand @TonyM1958 correctly, the V2 batteries have a maximum usable capacity of 80% of nominal, vs 90% of nominal for the V1? I wonder if this is being properly explained to prospective customers....

**Reply by TonyM1958** [reply]
Yes, that's correct - the usable capacity on v2 BMS and batteries appears to be 80% with a 10% min soc, as you loose 10% at the bottom end AND 10% at the top end as it only charges to 90% of capacity. For obvious reasons, Fox don't talk much about this. However, I'm happy with it filling to 90% if it means my battery charge / discharge cycles are more reliable and my batteries are likely to live for longer.

**Reply by FozzieUK** [reply]
If anyone has a v2 BMS / v2 only pack and is within 50 miles of manchester, and they don't mind me sitting in front of the pack for an hour to capture a few logs - i'd love to know how the batteries / bms reconciles all of that.

**Reply by calum-mcfarlane** [reply]
Feels like a bit of a hack to get around the fact that their BMS isn't very "nice" to the batteries, but seems either way you either give up a bit of capacity, or a bit of charging speed at the top end, or possibly a little of both. I'm thinking a MaxSoC in the winter of 95-97%, plus a bit of a taper on the charging current as @FozzieUK describes, might allow us to choose our own compromise between performance and longevity.

**Reply by TonyM1958** [reply]
@FozzieUK I'm down in Swindon unfortunately or you could drop over. I presume you are looking the capture the conversation on the CAN bus between the BMS and batteries?

@calum-mcfarlane I limit the max charge current to 13A, based on a force charge filling the batteries in the 3 hour Flux window. If you have v2 BMS and batteries, 100% max soc = 90% charge in any case, whether from solar or grid. It's just v1 that will batter the batteries until they get up to 100%...

@andrewmooreio just to try and validate the theory... if you had 2 x HV2600 v1 and 2 x HV2600 v2 and it's the batteries that are limiting charging to 100% and 90% respectively, then I would expect your 100% SoC to give you 2.56 + 2.56 + (2.56 + 2.56) * 0.9 = 9.73kwh - which seems to be about what you are actually seeing?

**Reply by andrewmooreio** [reply]
> 
> 
> > From what I have seen, this depends on the version of BMS you have - v1 tries to fill the batteries to 100% but v2 stops at 90%. So, if you have v2, you can expect the batteries to be full and SoC to read 100% when BMS_kWh_remaining = 10.24 * 0.9 = 9.2kWh. The fact you are getting 9.6 or 9.7kwh suggests to me that you have BMS v1 and there may be a cell balance issue that stops the BMS from getting to 10.24kWh?
> 
> 
> 
> @TonyM1958 the old V1 BMS used to charge to 99% but the latest firmware makes it the same as V2 BMS, both are now consistent and charge to 100%  - they both use same the same KwR remaining calculation.
> 
> 
> 
> @andrewmooreio It does sound like you have one or more packs slightly out of balance, the first check is do you have the latest V2 firmware installed? - BCU is 1.008, BMU is 1.07 (you can see these from the inverter panel select About, Batt Ver)
> 
> Each of the packs manages it's own cell balance, the BMS manages the pack balance - both of these are helped by occasionally running the pack to 10% and then recharging to 100%, preferably with the dc charge current set no higher than 5A which gives the balance circuits more time to do their job.

That would make sense. I'm travelling at the moment and can't verify for 2 weeks.

We had 2 batteries installed at the end of March. I then added an additional 2 in May. If they don't match then that'd be why.

**Reply by andrewmooreio** [reply]
Hi @TonyM1958 @FozzieUK. I've looked through every menu on the inverter and the only battery version info is in Menu > About > Bat Ver:

=== Bat Ver ===
M: 00.01.001
S1: 01.00.01.04
S2: 02.00.01.04
S3: 03.00.01.00
S4: 04.00.01.00

I've balanced the battery packs by discharging to 10% SoC and charging back to 100% at 5A. The max capacity hasn't really changed, it varies between 9.55 kWh and 9.75 kWh. The variation will be due to the SoC only being reported as an int, if there were decimals I guess it would be more accurate.

<img width="1459" alt="image" src="https://github.com/nathanmarlor/foxess_modbus/assets/20435317/f37d434d-4ae8-4fb9-b52d-6da1f66a90c2">

To get this value, I have the following template, just in case there is anything wrong with the logic.

```
{% set batt_remaining = states('sensor.bms_kwh_remaining') | float(default=0) %}
{% set batt_soc = states('sensor.battery_soc') | int(default=0) %}
{{ (batt_remaining / batt_soc * 100) | round (2, default=0) }}
```

**Reply by FozzieUK** [reply]
@andrewmooreio Hi Andrew, from what Tony has said it seems the V2 only reports up to 90% of the available battery energy so your 9.55-9.75kWh would appear to be quite a healthy figure. 

As you have V2, your versions effectively mean M(aster) =BCU , S(lave) = BMU and the latest are BCU 1.008 and BMU 1.07 so you are quite a few versions behind. Much has been done on alignment with mixed V1 packs so I would really like to see what happens on the latest version of firmware as I would hope they have brought V2 into line with the way V1 reports (100%). 
At this stage I would recommend that you get in touch with your installer or drop Foxess an email and ask them to bring your battery versions brought up to date, they have been stable for over 2 months now.

**Reply by andrewmooreio** [reply]
Thanks @FozzieUK. I'll get them to update the firmware. While I'm at it, I'll get the inverter updated too. Do you know the latest H1 firmware version?

**Reply by FozzieUK** [reply]
Ok, let us know how you get on 👍  - the latest H1 Master is 1.60 and H1 Manager is 1.63

**Reply by TonyM1958** [reply]
Just a quick check - how balanced is your array and inverter? A number of people have reported DC Bus Volt too High recently where solar kwp significantly exceeds their inverter kW. No evidence to support this but all the people that reported it seem to be on the latest inverter firmware. I'm reasonably balanced with 6.4kwp into 6kw and was planning to update mine from Manager 1.57 but thought I'd leave it a month or so while Fox investigate this issue, just in case... may be nothing but my inverter seems to be working fine just now.

**Reply by andrewmooreio** [reply]
I've asked FoxESS to update the firmware of all devices, and they said they've done it, yet only the BMU seems to have been done.

New values:

=== Bat Ver ===
M: 00.01.008
S1: 01.00.01.04
S2: 02.00.01.04
S3: 03.00.01.00
S4: 04.00.01.00

Max capacity has jumped up - is this just because the BMU has lost the previous data has to re-learn?

![image](https://github.com/nathanmarlor/foxess_modbus/assets/20435317/c88a5f6f-0493-4a66-bc1d-8ebfa160bda9)

**Reply by FozzieUK** [reply]
That's interesting, the slaves are different versions - they have to do master first and sometimes slave get interrupted by a timeout - best to leave it overnight and if it hasn't updated by tomorrow send them your mismatched Bat Ver's and ask them to check.

Their tech support said that when the firmware is upgraded it loses all the counters and has to re-learn the pack again.

**Reply by andrewmooreio** [reply]
I was impatient. Fox emailed an hour or so ago to say the updates were completed.

My inverter went offline about 15 minutes ago so I went to have a look. It had "Updating..." on the screen. I let it finish and now the battery packs are up to date. 

Guess it just takes a while.

**Reply by TonyM1958** [reply]
Please can you post all the versions of the inverter and batteries once the update is complete? Just useful to keep a track of what versions they actually updated you to. Also very interested in where the capacity ends up once the batteries have settled. Was hoping it would improve on the 90% I've been getting on BMS and battery v2...

**Reply by FozzieUK** [reply]
It’s packet based and slow to update, the master/manager on their own can take over 30 minutes and then the inverter has to restart, the batteries are usually a bit quicker but if you are a behind on your f/w versions sometimes it needs several sequential updates. 
Be really interested to see how the kwh remaining behaves after a couple of charge cycles 👍

**Reply by andrewmooreio** [reply]
Now I'm on the latest firmware for the inverter, BCU and BMU. I've rebalanced the batteries by discharging to 10% and then fully charging to 100% at 5A.

The behaviour is unexpected, my max capacity has increased fairly significantly. From ~9.6 kWh to ~10.3 kWh, which is now worryingly high.

![image](https://github.com/nathanmarlor/foxess_modbus/assets/20435317/99270efd-c549-4a25-9bcb-a4e3fc589d77)

1st Aug 16:00 is when the firmware upgrade took place, and you can see the jump in capacity. The battery was full at this time.

I then drained the battery to 10%, hitting 10% on 3rd Aug 08:00. The max capacity then went a bit haywire whilst charging back up to 100%, finishing on 4th Aug 11:30.

It may take more charge/discharge cycles to settle.

**Reply by FozzieUK** [reply]
Great stuff, I was hoping this would be the case for V2 systems, the older V1 packs have always reported this way and I think much of the 'imbalance' people experienced late last year / early this year with mixed V2/V1 packs was around differently reported SoC's and kWh remaining metrics, those systems have all settled down on the latest firmware.

It looks like your V2 system using the latest firmware is reporting battery kWh remaining as you would expect, which is very good news, i'll keep my fingers crossed 👍

**Comment by TonyM1958** [comment]
BTW - just in case this is firmware dependent - here's what versions I get reported (via HA):

![image](https://github.com/nathanmarlor/foxess_modbus/assets/63789168/a0a3892d-0588-4521-bc91-2acafa11ec6b)

Not worked out what this means yet, but HA also reports my BMS Module Type as 84 with 6 subsystems (batteries)?

![image](https://github.com/nathanmarlor/foxess_modbus/assets/63789168/84903d80-2866-4f44-9abc-933c7b789032)

**Reply by FozzieUK** [reply]
@TonyM1958 Thank - yes i’ve captured and part documented the Canbus between Inverter and Battery, and the RS485 between BMS and batteries for HV2600 V1, i’m just starting to look for other battery combinations so I can document the differences.

The subsystem number is the number of packs fitted (my 11047 shows 8), module type is the battery hardware fitted HV2600 V1 reports type 82 and so it looks like 84 must be HV2600 V2 - thanks that’s one for my notes.

Your inverter firmware is good, it’s not the very latest but a solid combination, the battery firmware are early versions, however most of the recent changes have been to support v1 mixed packs. With no release notes it’s a guess as to what else is in, word of mouth is they also improved temperature compensation and balance so I would be inclined to get them updated before winter if possible.

**Reply by TonyM1958** [reply]
I keep a list of registers and decodes in the spread sheet here: https://github.com/TonyM1958/HA-FoxESS-Modbus/wiki/Modbus-Map.xls

The other BMS Module Type value I've seen is 83 - this is the ECS4100 v1 master. Also worth noting the BMS Subsytem Number appears to be byte encoded, with the high byte describing the battery type and the low byte the number of batteries. For example, I've seen HV2600 = 0, ECS4100 = 7. It would be useful if we could scan some other battery types to see what other BMS and Battery codes are out there.

When I asked Fox a few months ago about my BMS and battery firmware, they told me this was the lastest version for HV BMS v2 abd HV2600 v2. I know that there are different firmware and battery versions for HV BMS v1 and HV2600 v1. Would be good to confirm the versions you mentioned above are for BMS v2 and not just for BMS v1?

**Reply by FozzieUK** [reply]
@TonyM1958 ok, thanks - i’ll update the notes with ECS4100 v1 as type 83 👍 

The BMS sends the number of packs as a single byte in a status message each poll cycle, but the battery type comes from a pack message with each pack being incremented on every poll - it can take a number of polls to get through all the packs. 
I can imagine if they encode number of packs and type together, there would be occasions (startup being one) where the register would be part populated, the number of packs would be available within 30 seconds but it would be a minute or two before it gets the battery type as the BMS has to complete it’s pack registration / startup process first, then poll all the packs  before it can update the inverter.

I just double checked the battery firmware releases, these are latest for v1 and v2 BMS versions - there was quite a bit of firmware update activity every few weeks over March/April, but these latest releases have been stable for a while (the v1 versions were only updated in June to bring the versions into line)

![IMG_1430](https://github.com/nathanmarlor/foxess_modbus/assets/113460294/40a66632-0cc2-4679-a1f0-44e919cd6d3d)
